kartothek/serialization/_parquet.py

```diff
@@ -175,6 +189,52 @@ def restore_dataframe(
         else:
             return df
 
+    @classmethod
+    def restore_dataframe(
+        cls,
+        store: KeyValueStore,
+        key: str,
+        filter_query: Optional[str] = None,
+        columns: Optional[Iterable[str]] = None,
+        predicate_pushdown_to_io: bool = True,
+        categories: Optional[Iterable[str]] = None,
```

Comment:
I think you need stronger guaranteed here than Iterable; after all, the first try to consume those iterators, leaving wrong arguments on the second try.

Reply:
Probably `Sequence` is then a better candidate. It is very likely that this is not 100% properly done in many places.

Imho, it's not entirely clear, though. afaik, the difference between `Sequence` and `Iterable` is that a sequence must define `__len__` and `__getitem__` but that would exclude sets.

From your comment I would judge we need something like "iterable but not generator" but I do not know which type that should be

Reply:
If you really need to support sets and lists, I think the right type is `Collection[str]`. OTOH, why not simply use something concrete here that reflects how it's used?

For instance, `categories` needs to be a list already, as it's passed to an external function expecting a list.
`columns` can;t be a set today because some functions call `columns[:]` to copy it, which does not work for sets ...

Reply:
If it is that explicit, I would argue it is even mandatory.
Using a sequence/iterable instead of explicit list is a recommendation of mypy since, e.g. sequence is covariant in its variables where a list isn't. It also makes it very explicit that the object is not mutated and allows for more compatible typing. 

Some documentation about this can be found here https://mypy.readthedocs.io/en/latest/generics.html#variance-of-generic-types 

That all said, I'm fine either way. If mypy doesn't complain, it's good enough for me

Reply:
For my own curiosity

```python
from collections import Iterable, Sequence
In [4]: isinstance([1, 2], Sequence)
Out[4]: True

In [5]: isinstance((1, 2), Sequence)
Out[5]: True

In [6]: isinstance((1, 2), Iterable)
Out[6]: True

In [7]: isinstance({1, 2}, Iterable)
Out[7]: True

In [8]: isinstance({1, 2}, Sequence)
Out[8]: False

In [9]: isinstance((x for x in [1, 2]), Iterable)
Out[9]: True

In [10]: isinstance((x for x in [1, 2]), Sequence)
Out[10]: False
```

generators are not sequences

Reply:
In my personal opinion, this should be `set[str]`, the logical data type but this would break the tests and is not really the "paradigm" we're using here anyways. I'd also rather have the type be explicit than duck typing.

I just copied the original function signature, we can probably have this debate somewhere else

Reply:
As I said, set is wrong.
Why not just change it to list?